syntax = "proto3";

package post_service;

service PostService {
    rpc CreatePost (CreatePostRequest) returns (CreatePostResponse);
    rpc UpdatePost (UpdatePostRequest) returns (UpdatePostResponse);
    rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
    rpc GetPost (GetPostRequest) returns (GetPostResponse);
    rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}

message Post {
    string id = 1;
    string title = 2;
    string content = 3;
    string user_id = 4;
    string created_at = 5;
    string updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message CreatePostRequest {
    string title = 1;
    string content = 2;
    string user_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message CreatePostResponse {
    string id = 1;
}

message UpdatePostRequest {
    string id = 1;
    string title = 2;
    string content = 3;
    string user_id = 4;
    bool is_private = 5;
    repeated string tags = 6;
}

message UpdatePostResponse {}

message DeletePostRequest {
    string id = 1;
    string user_id = 2;
}

message DeletePostResponse {}

message GetPostRequest {
    string id = 1;
    string user_id = 2;
}

message GetPostResponse {
    Post post = 1;
}

message ListPostsRequest {
    string user_id = 1;
    int32 page = 2;
    int32 per_page = 3;
}

message ListPostsResponse {
    repeated Post posts = 1;
    int32 total = 2;
}